### PR TITLE
nix: Build drun from public repository

### LINKED ISSUE #2567: CI/Buildsystem: drun for all
Currently, all jobs that need `drun` (some of the tests, but also the profiling reports and performance comparison) cannot be run by external contributors or our Github based CI. We want this to change.

The most likely way forward is to
 * do what’s necessary internally to get the `drun` code included in the code dump in https://github.com/dfinity/ic
 * include it from there, with a nix setup (using plain `pkgs.rustPlatform.buildRustPackage`, we don't need finer build granularity for this) on our side

### PR DESCRIPTION
**Pull request by nomeata**
The Motoko test suite uses a tool called `drun`, which is an Internet Computer emulator in a single program. Previously it was only available internally within DFINITY, but since https://github.com/dfinity/ic/commit/a0c8c9a1de828d0881eeedf8764cc4b11d5f0fb7 it’s part of the open source code dump.

So with this we can stop relying on internal repositories (fixes #2567). Also, no VPN needed anymore for the DFINITY employees! :-)

`drun` is built using plain `buildRustPackage`, i.e. as one bit nix deriation. This is ok for dependencies that _we_ don’t hack on and rarely change. And due to the nix cache, everybody will just download the `drun` binary, I hope.

### CODE REVIEW

**Comment by nomeata** [comment]
`drun` is a mild beast to build. It worked in `nix-shell`, but properly nixified I am currently hitting
```
error: failed to run custom build command for `lmdb-rkv-sys v0.11.99 (https://github.com/dfinity-lab/lmdb-rs?rev=1cf86b5cc09947e94a787065cadd163a42ef7f18#1cf86b5c)`

Caused by:
  process didn't exit successfully: `/build/rs/target/release/build/lmdb-rkv-sys-e417a733519aec1b/build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=bindgen.rs
  cargo:rerun-if-changed=src/lib.rs

  --- stderr
  error: header '/build/drun-vendor.tar.gz/lmdb-rkv-sys/lmdb/libraries/liblmdb/lmdb.h' does not exist.
  thread 'main' panicked at 'Unable to generate bindings: ()', /build/drun-vendor.tar.gz/lmdb-rkv-sys/bindgen.rs:77:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Might be a problem with the vendoring of `buildRustPackage`

**Comment by nomeata** [comment]
Ok, it works on linux now. On Darwin I get
```
   Compiling ic-protobuf v0.8.0 (/private/tmp/nix-build-drun.drv-0/rs/protobuf)
error: linking with `/nix/store/d4p8v32sivk3zj3jklx3irswbwwigxcd-clang-wrapper-7.1.0/bin/cc` failed: exit code: 1
  |
  = note: "/nix/store/d4p8v32sivk3zj3jklx3irswbwwigxcd-clang-wrapper-7.1.0/bin/cc" "-m64" "-arch" "x86_64" "-L" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.0.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.1.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.10.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.11.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.12.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.13.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.14.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.15.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.2.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.3.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.4.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.5.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.6.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.7.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.8.rcgu.o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.build_script_build.b3r89hap-cgu.9.rcgu.o" "-o" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04" "/private/tmp/nix-build-drun.drv-0/rs/target/release/build/ic-protobuf-0520392e6e1bdc04/build_script_build-0520392e6e1bdc04.2ecyg4gj2nq5q8sn.rcgu.o" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps" "-L" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libprost_build-9e4931c037e5f90a.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libtempfile-3beb6cbf4b61c026.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/librand-6bf05eb60c328fd8.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/librand_chacha-362eaeadbb76cb89.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libppv_lite86-e9598d000bd4c1dc.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/librand_core-2c5a7b5d87ede514.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libgetrandom-61cad2cdc3c03a99.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/liblibc-e49ab4a52b60d1a9.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libremove_dir_all-4aeb96e396c732f9.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libcfg_if-af390cb2c34a7d9c.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libpetgraph-7904fb661438e612.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libindexmap-0b099c562b418e2a.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libhashbrown-4fb77f3e0c9a45a4.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libserde-496a02014772cf82.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libfixedbitset-43164489e8f5b471.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libheck-ec9b581f67411a5c.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libunicode_segmentation-f17934c5317c29f0.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libmultimap-bbae456618b5c70a.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/liblog-4f794ecc0881e887.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libcfg_if-8b156c9e8dc44909.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libitertools-d6169400f7ca11e3.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libeither-78883f98dd43192c.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libprost_types-bb0ae4f03f330843.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libprost-cf77f28a60793646.rlib" "/private/tmp/nix-build-drun.drv-0/rs/target/release/deps/libbytes-956811e24b957e79.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libstd-8ecebea30c961804.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-78b8371c8829edb9.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libobject-11c2250f76ee79de.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libaddr2line-138200afe481d69a.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libgimli-50edf7ed96dca7d7.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-f5add760ba32d8c5.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-107795e7ae43b1fc.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-1a28e1881ca37aeb.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libunwind-0b60cab5e184de42.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-20066734a370e7ce.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/liblibc-0688efbb4c57e170.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/liballoc-0c8da59b450dd54a.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-159337c2037ece59.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libcore-1ba7a3965a87fb70.rlib" "/nix/store/ly04c58zsvb4c219sv6q393dik249hjy-rustc-1.52.1/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-2f50c7e4a63f3402.rlib" "-framework" "Security" "-liconv" "-lSystem" "-lresolv" "-lc" "-lm" "-liconv"
  = note: ld: framework not found Security
          clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: aborting due to previous error

error: could not compile `ic-protobuf`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
error: builder for '/nix/store/7k6m8q7wck5rxydmdrzgffcqgzbcnq0a-drun.drv' failed with exit code 101;
       last 10 log lines:
       >           clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
       >           
       >
       > error: aborting due to previous error
       >
       > error: could not compile `ic-protobuf`
       >
       > To learn more, run the command again with --verbose.
       > warning: build failed, waiting for other jobs to finish...
       > error: build failed
```
and it will be no fun debugging that without access to a darwin machine. We’ll see, but not now.

**Comment by nomeata** [comment]
Ok, progress, `drun` built on OSX. Now hitting
```
selftail: [tc] [run] [run-ir] [run-low] [comp] [comp-ref] [valid] [valid-ref] [drun-run]
--- selftail.drun-run.ret (expected)
+++ selftail.drun-run.ret (actual)
@@ -0,0 +1 @@
+Return code 134
--- selftail.drun-run (expected)
+++ selftail.drun-run (actual)
@@ -1,4 +1,10 @@
 ingress Completed: Reply: 0x4449444c016c01b3c4b1f204680100010a00000000000000000101
 debug.print: ok1
 debug.print: ok2
-ingress Err: IC0502: Canister rwlgt-iiaaa-aaaaa-aaaaa-cai trapped: stack overflow
+
+thread 'MR Batch Processor' has overflowed its stack
+fatal runtime error: stack overflow
+/tmp/nix-build-test-drun-dbg.drv-0/test-run-drun-src/run-drun/../drun-wrapper.sh: line 49:  2787 Done                    ( echo "create"; echo "install $ID $1 0x"; if [ -n "$2" ]; then
+    LANG=C perl -ne 'print "$1 '$ID' $2\n" if m,^//CALL (ingress|query) (.*),;print "upgrade '$ID' '"$1"' 0x\n" if m,^//CALL upgrade,; ' $2;
+fi )
+      2789 Abort trap: 6           | drun -c "$CONFIG" --extra-batches $EXTRA_BATCHES /dev/stdin
```
Did anyone see that before?

**Comment by nomeata** [comment]
Fun, if I set `RUST_MIN_STACK`, it actually makes the Wasm stack larger, and now I get
```
-ingress Err: IC0502: Canister rwlgt-iiaaa-aaaaa-aaaaa-cai trapped: stack overflow
+debug.print: ok3 (unreachable on drun)
+ingress Completed: Reply: 0x4449444c0000
```
on linux. It’s pretty smelly that the size of the wasm stack depends on the rust stack here…

I think I'll just remove this part of the test. There is little value in us testing how `drun` behaves in a stack overflow situation.

**Comment by nomeata** [comment]
Reported sporadic segfault of `nix-build` on darwin as https://github.com/NixOS/nix/issues/5125

**Comment by nomeata** [comment]
Can someone in the in-group tell me the evaluation error at https://hydra.dfinity.systems/jobset/dfinity-ci-build/motoko.pr-2713#tabs-errors?

**Comment by nomeata** [comment]
> Can someone in the in-group tell me the evaluation error at https://hydra.dfinity.systems/jobset/dfinity-ci-build/motoko.pr-2713#tabs-errors?

Ah, probably 
```
error: attribute 'perf' missing, at /home/jojo/dfinity/motoko/ci-pr.nix:66:22
```
that should be temporary only, and go away again when `master` has the changes of this PR.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -5,20 +5,11 @@ jobs:
...
-      with:
-        # We need a nix version with NixOS/nix#3564 to be able to use
-        # `--max-jobs`, so using NixOS/nix@323e545. See 8365e73 for the problem
-        # with `--max-jobs`. Install URL is obtained from CI logs for the nix
-        # commit: https://github.com/NixOS/nix/runs/2895674848 expand "Run
-        # cachenix/install-nix-action@v13" job, then the "Run
-        # cachenix/install-nix-action@v13" step, and you'll see a line starting
-        # with `install_url`.
-        install_url: https://nixos-nix-install-tests.cachix.org/serve/i6laym9jw3wg9mw6ncyrk6gjx4l34vvx/install
-        install_options: '--tarball-url-prefix https://nixos-nix-install-tests.cachix.org/serve'
```
Comment (osa1) [inline .github/workflows/test.yml:21]: Why is this removed?

**Comment by osa1** [comment]
> For reasons I don’t understand the error message on stack overflow on darwin changed. I removed that test.

Changed how? As much as I hate the sed lines in `run.sh` (actually, the whole `run.sh`), perhaps we should "normalize" the test outputs instead of disabling the test.

File: .github/workflows/test.yml (hunk 1 above)
Comment (nomeata) [reply .github/workflows/test.yml:21]: I am still experimenting. `nix` master segfauls on darwin, the released version doesn’t. But then we are of course hitting
```
error: all build users are currently in use; consider creating additional users and adding them to the 'nixbld' group
```
again. I’ll see if there is a way to create more users.

**Comment by osa1** [comment]
FYI hydra error:

```
in job ‘all-jobs’:
error: --- EvalError --- hydra-eval-jobs
at: (66:22) in file: /nix/store/wgn1bzfbmdzg0lyp2zj6mpfvq3704gqc-source/ci-pr.nix

    65|       else
    66|         diff-stats ${baseJobs.tests.perf}/stats.csv ${prJobs.tests.perf}/stats.csv > $out/report;
      |                      ^
    67|       fi

attribute 'perf' missing

in job ‘perf-delta’:
error: --- EvalError --- hydra-eval-jobs
at: (66:22) in file: /nix/store/wgn1bzfbmdzg0lyp2zj6mpfvq3704gqc-source/ci-pr.nix

    65|       else
    66|         diff-stats ${baseJobs.tests.perf}/stats.csv ${prJobs.tests.perf}/stats.csv > $out/report;
      |                      ^
    67|       fi

attribute 'perf' missing
```

**Comment by nomeata** [comment]
> > For reasons I don’t understand the error message on stack overflow on darwin changed. I removed that test.
> 
> Changed how? As much as I hate the sed lines in `run.sh` (actually, the whole `run.sh`), perhaps we should "normalize" the test outputs instead of disabling the test.

See https://github.com/dfinity/motoko/pull/2713#issuecomment-897511631

Tests always have a friction/benefit trade-off. And chasing drun bugs or peculiarities of the `drun` build system is a distraction, so I am inclined not to get hung on this.

We could leave the test and only run it in `ic-ref-run`. What do you think?

**Comment by osa1** [comment]
> +thread 'MR Batch Processor' has overflowed its stack

I think this is a configuration error in replica. Execution team says this happens when we pass a larger stack size to wasmtime than the process allows. (gotta go now, will investigate later)

**Comment by nomeata** [comment]
> > +thread 'MR Batch Processor' has overflowed its stack
> 
> I think this is a configuration error in replica. Execution team says this happens when we pass a larger stack size to wasmtime than the process allows. (gotta go now, will investigate later)

Maybe. That was my assumption too. But when I increased the rust thread stack size, it seems to have made the wasm stack overflow go away (see comment https://github.com/dfinity/motoko/pull/2713#issuecomment-897541139 if you haven’t read it all yet). At that point I retracted, but maybe it is worth investigating and fixing.

**Comment by nomeata** [comment]
Yay, it’s green!

**Comment by nomeata** [comment]
Let’s see how well it works after pulling in latest `master`.

If it turns green, shall we move forward with this?

This modifies `.github` files, so need to squash merge using the button, not mergify, and thus manually edit the commit message. I can do that, once approved.

File: .github/workflows/test.yml (hunk 1 above)
Comment (osa1) [reply .github/workflows/test.yml:21]: Ping @nomeata

**Comment by osa1** [comment]
Would be good to make sure we won't have "nix users" issues as discussed above.

Also the problem with stack overflows can be solved easily I think. I will take a look tomorrow.

**Comment by nomeata** [comment]
> Would be good to make sure we won't have "nix users" issues as discussed above

As long as we don't set `-j` to more than 4 (number of nix users), it seems to work fine. And more than 4 times is probably not useful on these free runners there. So that shouldn't hold us up.

As for drun having not consistent error messages: maybe worth investigating, but unless it's trivial I'd not wait for it. Especially if it would require waiting for a new source code dump…

**Comment by osa1** [comment]
> but unless it's trivial I'd not wait for it. Especially if it would require waiting for a new source code dump…

I think we just need a configuration file to specify heap size, or pass a parameter. I will check.

**Comment by osa1** [comment]
@nomeata setting env variable `RUST_MIN_STACK=8192` should fix it

**Comment by nomeata** [comment]
> > but unless it's trivial I'd not wait for it. Especially if it would require waiting for a new source code dump…
> 
> I think we just need a configuration file to specify stack size, or pass a parameter. I will check.

rust stack or wasm stack?

As I said above, I changed rust stack size and it oddly hid the wasm stack over flow. Changing wasm stack size would change platform behavior, so surely not the right way.

**Comment by nomeata** [comment]
Did you see my comment in https://github.com/dfinity/motoko/pull/2713#issuecomment-897541139

**Comment by osa1** [comment]
> As I said above, I changed rust stack size and it oddly hid the wasm stack over flow. Changing wasm stack size would change platform behavior, so surely not the right way.

That's really weird. I'm talking to the execution team. We have 3 notions of stack:

1. Rust process stack
2. Wasm stack (manipulated by push/pop instructions)
3. Stack allocated in Wasm heap, maintained by generated code

Execution team says there's a bug in the public release that causes overflowing (1) and the `RUST_STACKSIZE` value I showed above should fix it. I think setting `RUST_STACKSIZE` is a good idea regardless as it works around a replica bug in the public release.

Once that's done you say (2) overflows. We discussed how this can happen, and exec team says they updated wasmtime recently (shipped with the public release) and the new version may be utilizing process's stack space more efficiently (for example, less spills in JIT generated code). If that's true then increasing the recursion amount should bring back the old output (trap).

**Comment by nomeata** [comment]
Thanks for the research! Annoying to have differences between the arches. Maybe we find parameters that work though.

**Comment by osa1** [comment]
> Annoying to have differences between the arches

I think this is unavoidable, unless we decide to not try to break any limits/ overflow anything. The problem is a JIT compiler will generate different code on x86 than ARM, and these codes will use the process's stack differently. Same for memory usage.

It should be fine within some resource limits though. Perhaps a good motivation to avoid overflowing the stack or other limits in tests?

**Comment by nomeata** [comment]
> . Perhaps a good motivation to avoid overflowing the stack or other limits in tests?

… which is why in this PR I suggest to avoid overflowing the stack in tests? Now I am confused what we are discussing here…

BTW, the same problem has prevented auto-updating the `dfinity` pin in https://github.com/dfinity/motoko/pull/2699, so not really related to the nix changes here, merely to some change in execution.

**Comment by nomeata** [comment]
This seems to work now

**Comment by osa1** [comment]
What happened to the "all build users are currently in use" issue? It should happen again I think, we weren't able to just increase jobs before without bumping nix version.

**Comment by nomeata** [comment]
As long as we don't set -j to more than 4 (number of nix users), it seems to work fine. And more than 4 times is probably not useful on these free runners there. So that shouldn't hold us up.

**Review by osa1 — APPROVED** [review]

**Comment by nomeata** [comment]
Thanks for the suggestion of bumping both stack limit _and_ recursion depth.

**Pull request merged**